zebra-network/src/meta_addr.rs

```diff
@@ -103,51 +410,407 @@ impl PartialOrd for PeerAddrState {
+                untrusted_last_seen,
+            } if !has_been_attempted => Some(MetaAddr::new(
+                addr,
+                NeverAttemptedGossiped {
+                    untrusted_services,
+                    // Keep the first time we got
+                    untrusted_last_seen: old_untrusted_last_seen.unwrap_or(untrusted_last_seen),
+                },
+            )),
+            NewAlternate {
+                addr,
+                untrusted_services,
+            } if !has_been_attempted => Some(MetaAddr::new(
+                addr,
+                NeverAttemptedAlternate {
+                    untrusted_services,
+                    // Keep the first time we got
+                    untrusted_last_seen: old_untrusted_last_seen.unwrap_or_else(Utc::now),
+                },
+            )),
+
+            // New entry, but already existing (seed) or attempted (others):
+            // Skip the update entirely
+            NewSeed { .. } | NewGossiped { .. } | NewAlternate { .. } => None,
+
+            // Attempt:
+            // Update last_attempt
+            UpdateAttempt { addr } => Some(MetaAddr::new(
+                addr,
+                AttemptPending {
+                    last_attempt: Utc::now(),
+                    last_success: old_last_success,
+                    last_failed: old_last_failed,
+                    untrusted_last_seen: old_untrusted_last_seen,
+                    untrusted_services: old_untrusted_services,
+                },
+            )),
+
+            // Responded:
+            // Update last_success and services
+            UpdateResponded { addr, services } => Some(MetaAddr::new(
+                addr,
+                Responded {
+                    // When the attempt message arrives, it will replace this default.
+                    // (But it's a reasonable default anyway.)
+                    last_attempt: old_last_attempt.unwrap_or_else(Utc::now),
+                    last_success: Utc::now(),
+                    last_failed: old_last_failed,
+                    untrusted_last_seen: old_untrusted_last_seen,
+                    services,
+                },
+            )),
+
+            // Failed:
+            // Update last_failed and services if present
+            UpdateFailed { addr, services } => Some(MetaAddr::new(
+                addr,
+                Failed {
+                    // When the attempt message arrives, it will replace this default.
+                    // (But it's a reasonable default anyway.)
+                    last_attempt: old_last_attempt.unwrap_or_else(Utc::now),
+                    last_success: old_last_success,
+                    last_failed: Utc::now(),
+                    untrusted_last_seen: old_untrusted_last_seen,
+                    // replace old services with new services if present
+                    untrusted_services: services.or(old_untrusted_services),
+                },
+            )),
```

Comment:
Security: Zebra should stop gossiping failure times as last_seen times #1868

Part 1a: Keep the previous last seen time for failed peers